### PR TITLE
Allow default network mode, as required for windows containers.

### PR DESCRIPTION
**Pull request by ksagle77**
##### SUMMARY

Windows containers require the <default> network mode. In boto3, this is configured by not passing a network_mode paramter, but this mode is not supported by this module.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

cloud

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by xp0** [comment]
This would be really nice to have in the next release

**Comment by tremble** [comment]
@xp0 @ksagle77 

https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html seems to show when not set it'll default to 'bridge' which is what this module also does.  Could you provide some information to show when / why this is necessary?

**Comment by ksagle77** [comment]
@tremble 

Hey Mark,

The issue is that docker for windows has it's own networking mode (NAT), which requires that no network mode is selected. Choose bridge causes it to fail. From https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html:

> Docker for Windows uses a different network mode (known as NAT) than Docker for Linux. When you register a task definition with Windows containers, you must not specify a network mode. If you use the AWS Management Console to register a task definition with Windows containers, you must choose the default network mode.

Kevin

File: lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py [hunk 1]
```diff
@@ -62,7 +62,7 @@
             - C(awsvpc) mode was added in Ansible 2.5
         required: false
         default: bridge
-        choices: [ 'bridge', 'host', 'none', 'awsvpc' ]
+        choices: [ 'default', 'bridge', 'host', 'none', 'awsvpc' ]
```
Comment (tremble) [inline lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py:67]: I'd like to see the effect of 'default' documented in the description above.

**Review by tremble — CHANGES_REQUESTED** [review]
A couple of minor things here.

1. Could you add a changelog fragment https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs
2. Please add something to the description to explain what 'default' does

**Comment by ksagle77** [comment]
@tremble 

Hey Mark,

I've added the change fragment and a description. Let me know if it should be more detailed.

Kevin

File: lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py [hunk 2]
```diff
@@ -60,9 +60,11 @@
         description:
             - The Docker networking mode to use for the containers in the task.
             - C(awsvpc) mode was added in Ansible 2.5
+            - Windows containers must use default mode, which will utilize docker NAT networking. Selecting this for a linux container will
```
Comment (tremble) [inline lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py]:
```suggestion
            - Windows containers must use I(network_mode=default), which will utilize docker NAT networking.
            - Setting I(network_mode=default) for a Linux container will use bridge mode.
```

File: changelogs/fragments/59597-ecs-allow_default_network_mode.yml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - ecs_task_definition - Add use of default network mode as required by windows ECS tasks.
```
Comment (tremble) [inline changelogs/fragments/59597-ecs-allow_default_network_mode.yml]:
```suggestion
  - ecs_task_definition - Add ``network_mode=default`` to support Windows ECS tasks.
```

**Review by tremble — CHANGES_REQUESTED** [review]
Some tiny tweaks in the wording, but generally looking good to me.

File: lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py [hunk 4]
```diff
@@ -60,8 +60,8 @@
         description:
             - The Docker networking mode to use for the containers in the task.
             - C(awsvpc) mode was added in Ansible 2.5
-            - Windows containers must use I(network_mode=default), which will utilize docker NAT networking.
-            - Setting I(network_mode=default) for a Linux container will use bridge mode.
+            - Windows containers must use network_mode=default, which will utilize docker NAT networking.
```
Comment (tremble) [inline lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py]:
```suggestion
            - Windows containers must use I(network_mode=default), which will utilize docker NAT networking.
```

Using I(...) adds some extra highlighting that says "Option network_mode set to value 'default'"

File: lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py [hunk 5]
```diff
@@ -298,7 +302,7 @@ def main():
         revision=dict(required=False, type='int'),
         force_create=dict(required=False, default=False, type='bool'),
         containers=dict(required=False, type='list'),
-        network_mode=dict(required=False, default='bridge', choices=['bridge', 'host', 'none', 'awsvpc'], type='str'),
+        network_mode=dict(required=False, default='bridge', choices=['default','bridge', 'host', 'none', 'awsvpc'], type='str'),
```
Comment (tremble) [inline lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py]:
```suggestion
        network_mode=dict(required=False, default='bridge', choices=['default', 'bridge', 'host', 'none', 'awsvpc'], type='str'),
```

**Review by tremble — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py (hunk 5 above)
Comment (ksagle77) [reply lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py]: Oh, didn't realize that. Adding it back, thanks.

File: lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py [hunk 6]
```diff
@@ -60,9 +60,12 @@
         description:
             - The Docker networking mode to use for the containers in the task.
             - C(awsvpc) mode was added in Ansible 2.5
+            - Windows containers must use I(network_mode=default), which will utilize docker NAT networking.
+            - Setting network_mode=default for a Linux container will use bridge mode.
```
Comment (tremble) [inline lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py]:
```suggestion
            - Setting I(network_mode=default) for a Linux container will use bridge mode.
```

File: lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py [hunk 7]
```diff
@@ -60,9 +60,12 @@
         description:
             - The Docker networking mode to use for the containers in the task.
             - C(awsvpc) mode was added in Ansible 2.5
+            - Windows containers must use I(network_mode=default), which will utilize docker NAT networking.
+            - Setting I(network_mode=default) for a Linux container will use bridge mode.
+              use bridge mode.
```
Comment (tremble) [inline lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py]:
```suggestion
```

**Comment by tremble** [comment]
One last change (removing the duplicate text, unfortunately GitHub doesn't let you make suggestions against multiple lines yet) and then I *think* we're good to go.

**Comment by ksagle77** [comment]
Removed. Thanks for the help.

**Review by tremble — APPROVED** [review]

**Review by jillr — APPROVED** [review]
LGTM, tests pass locally, thanks very much for this patch @ksagle77!

**Comment by tremble** [comment]
LGTM

**Comment by tremble** [comment]
bot_status

**Pull request merged**